src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepLauncher.kt

```diff
@@ -24,22 +25,33 @@ import org.wfanet.panelmatch.client.launcher.InvalidExchangeStepException.Failur
 class ExchangeStepLauncher(
   private val apiClient: ApiClient,
   private val validator: ExchangeStepValidator,
-  private val jobLauncher: JobLauncher
+  private val jobLauncher: JobLauncher,
+  maxParallelExchangeSteps: Int? = null
 ) {
 
+  private val semaphore = if (maxParallelExchangeSteps !== null) Semaphore(maxParallelExchangeSteps) else null
+
   /**
    * Finds a single ready Exchange Step and starts executing. If an Exchange Step is found,
    * validates it, and starts executing. If not found simply returns.
    */
   suspend fun findAndRunExchangeStep() {
-    val (exchangeStep, attemptKey) = apiClient.claimExchangeStep() ?: return
+    if (semaphore !== null) semaphore.acquire()
```

Comment:
My one question here after reading more in kotlin docs is: should this be tryAcquire? Otherwise we'll queue up suspended requests much faster than we can clear them. The throttler should control the loop speed so we don't hammer this like a spinlock.